add endpoints to batch apply scripts and integrate with CLI

roperzh opened the pull request:
for #9828

Note: error messages doesn't match exactly the Figma copy because there are some inconsistencies between this API and the rest that I'd like to sanity check with product.

# Checklist for submitter

If some of the following don't apply, delete the relevant line.

- [x] Documented any API changes (docs/Using-Fleet/REST-API.md or docs/Contributing/API-for-contributors.md)
- [x] Input data is properly validated, `SELECT *` is avoided, SQL injection is prevented (using placeholders for values in statements)
- [x] Added/updated tests
- [x] Manual QA for all new/changed functionality

gillespi314 commented on ee/server/service/scripts.go:
```diff
@@ -383,3 +376,67 @@ func (svc *Service) GetHostScriptDetails(ctx context.Context, hostID uint, opt f
+		}
+
+		if byName[script.Name] {
+			// TODO: check error message
```
Can this TODO be removed?

gillespi314 commented on server/datastore/mysql/migrations/tables/migration.go:
```diff
@@ -49,6 +50,23 @@ WHERE
 	return count > 0
 }
 
+func indexExists(tx *sqlx.DB, table, index string) bool {
```
Nice!

gillespi314 commented on server/datastore/mysql/scripts.go:
```diff
@@ -321,3 +321,108 @@ WHERE
+	// existing ones if there's a match and no change
+	incomingNames := make([]string, len(scripts))
+	// at the same time, index the incoming scripts keyed by name for ease
+	// or processing
```
Nit: "or" -> "of"

gillespi314 commented on server/datastore/mysql/scripts.go:
```diff
@@ -321,3 +321,108 @@ WHERE
+		}
+
+		// insert the new scripts and the ones that have changed
+		for _, s := range incomingScripts {
```
Just double checking my understanding: We aren't comparing the script contents so the updated_at timestamp for all the kept scripts will be auto-updated. Is that the expected behavior?

roperzh replied:
your understanding is right, to be honest, I'm not sure.

I implemented this following the batch set profiles endpoint, which behaves similarly. I also noticed that the UI is using `created_at` to display timestamps for scripts.

gillespi314 commented on server/datastore/mysql/scripts_test.go:
```diff
@@ -517,3 +518,102 @@ VALUES
+
+		// compare only the fields we care about
+		m := make(map[string]uint)
+		for _, gotp := range got {
```
Nit: "gotp" -> "gotS" or "s" (I think "p" is because was based on the profiles tests?)

gillespi314 commented on server/service/integration_enterprise_test.go:
```diff
@@ -4906,3 +4907,109 @@ func (s *integrationEnterpriseTestSuite) TestApplyTeamsScripts() {
+			teamActivity = fmt.Sprintf(`{"team_id": %d, "team_name": %q}`, team.ID, team.Name)
+		}
+
+		// create an check activities
```
Nit: "an" -> "and"

gillespi314 commented on ee/server/service/scripts.go:
```diff
@@ -383,3 +376,67 @@ func (svc *Service) GetHostScriptDetails(ctx context.Context, hostID uint, opt f
 	return svc.ds.GetHostScriptDetails(ctx, h.ID, h.TeamID, opt)
 }
+
+func (svc *Service) BatchSetScripts(ctx context.Context, maybeTmID *uint, maybeTmName *string, payloads []fleet.ScriptPayload, dryRun bool) error {
```
Should we add the usual auth tests for the new endpoint?

roperzh replied:
I tried, but I quickly realized why none of the `ee/` service methods has them! there's an import cycle we need to solve in order to use the logic that sets ups the test suite that makes this test possible. 😞

gillespi314 reviewed (approved):
Looks great! Just a few minor questions and nits. 

I'll go ahead and approve in case you would rather address those later.

roperzh commented:
@gillespi314 thanks so much for the review! I addressed the comments. I think you're right in regards to the test but seems to be a general deficiency of the `/ee` service methods.

I will give it a try and follow up if I find a quick solution. Merging this to get ahead of the freeze.
